pkg/planner/core/logical_plan_builder.go

```diff
@@ -5266,76 +5268,143 @@ func buildColumns2HandleWithWrtiableColumns(
+// And it will try to prune columns that not used in pk or indexes.
+func pruneAndBuildSingleTableColPosInfoForDelete(
 	names []*types.FieldName,
-	handleCol util.HandleCols,
 	deletableIdxs []table.Index,
 	deletableCols []*table.Column,
 	tblInfo *model.TableInfo,
-) (TblColPosInfo, error) {
+	colPosInfo *TblColPosInfo,
+	prePrunedCount int,
+	nonPrunedSet *bitset.BitSet,
+) (int, error) {
 	// Columns can be seen by DELETE are the deletable columns.
 	tblLen := len(deletableCols)
-	offset, err := getTableOffset(names, names[handleCol.GetCol(0).Index])
-	if err != nil {
-		return TblColPosInfo{}, err
+
+	// If it's partitioned table, or has foreign keys, or is point get plan, we can't prune the columns, currently.
+	if tblInfo.GetPartitionInfo() != nil || len(tblInfo.ForeignKeys) > 0 || nonPrunedSet == nil {
+		indexColMap := make(map[int64]table.IndexRowLayoutOption, len(deletableIdxs))
+		for _, idx := range deletableIdxs {
+			idxCols := idx.Meta().Columns
+			colPos := make([]int, 0, len(idxCols))
+			for _, col := range idxCols {
+				colPos = append(colPos, col.Offset)
+			}
+			indexColMap[idx.Meta().ID] = colPos
+		}
+		colPosInfo.End = colPosInfo.Start + tblLen
+		colPosInfo.IndexesForDelete = indexColMap
+
+		return prePrunedCount, nil
 	}
-	end := offset + tblLen
-
-	// Index only records its columns' offsets in the deletableCols(the whole columns).
-	// So we need to first change the offsets to the real position of SELECT's output.
-	offsetMap := make(map[int]int, len(deletableCols))
-	// For multi-delete case, the offset is recorded for the mixed row. We need to use its position in single table's row layout.
-	//	e.g. The multi-delete case is [t1.a, t1.b, t2.a, t2.b] and There's a index [t2.b].
-	//	     The idxCols' original offset is [3]. And we should use [1] instead.
-	for i, name := range names[offset:end] {
-		found := false
-		for j, col := range deletableCols {
-			if col.Name.L == name.ColName.L {
-				offsetMap[j] = i
-				found = true
-				break
+
+	// Fix the start position of the columns.
+	originalStart := colPosInfo.Start
+	colPosInfo.Start -= prePrunedCount
+
+	// Mark the columns in handle.
+	visitedCols := make(map[int]struct{}, tblLen)
+	for i := 0; i < colPosInfo.HandleCols.NumCols(); i++ {
+		col := colPosInfo.HandleCols.GetCol(i)
+		visitedCols[col.Index-originalStart] = struct{}{}
+	}
+
+	// Mark the columns in indexes.
+	for _, idx := range deletableIdxs {
+		for _, col := range idx.Meta().Columns {
+			if col.Offset >= len(names) || deletableCols[col.Offset].Name.L != names[col.Offset].ColName.L {
+				return 0, plannererrors.ErrDeleteNotFoundColumn.GenWithStackByArgs(col.Name.O, tblInfo.Name.O)
 			}
+			visitedCols[col.Offset] = struct{}{}
 		}
-		if !found {
-			return TblColPosInfo{}, plannererrors.ErrDeleteNotFoundColumn.GenWithStackByArgs(name.ColName.O, tblInfo.Name.O)
+	}
+
+	// Fix the column offsets.
+	pruned := 0
+	fixedPos := make(map[int]int, len(deletableCols))
```

Comment:
Could the memory be reused for `visitedCols` and `fixedPos` in some way?  As it's a critical path for each statement.

Reply:
I merged the two maps into one. Now the memory pressure is the same as before.